akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -79,24 +79,21 @@ import com.github.ghik.silencer.silent
     new GraphStageLogic(shape) with OutHandler with InHandler {
       def decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
 
-      override def onPush(): Unit = {
+      override def preStart(): Unit = pull(in)
+      override def onPush(): Unit =
         try {
           val elem = grab(in)
-          if (p(elem)) {
-            push(out, elem)
-          } else {
-            pull(in)
-          }
+          if (p(elem)) emit(out, elem, () => if (!isClosed(in)) pull(in))
```

Comment:
performance of `filter` can be important and I'm thinking if we should avoid `emit` and instead have a manual `OptionVal` buffer in this stage?

Reply:
That was my gut feeling looking at this as well, emit will be the lambda + one `EmittingSingle` per element worst case?

Reply:
Made that change even if it's 20 lines more